src/table/iterator.rs

```diff
@@ -187,6 +200,261 @@ impl BlockIterator {
+            }
+            Err(err) => self.err = Some(err.into()),
+        }
+    }
+
+    fn seek_from(&mut self, key: &Bytes, whence: SeekPos) {
+        self.err = None;
+        if matches!(whence, SeekPos::Origin) {
+            self.reset();
+        }
+
+        let idx = util::search(self.table.as_ref().offsets_length(), |idx| {
+            use std::cmp::Ordering::*;
+            let block_offset = self.table.as_ref().offsets(idx).unwrap();
+            match COMPARATOR.compare_key(&block_offset.key, &key) {
+                Greater => true,
+                _ => false,
+            }
+        });
+
+        if idx == 0 {
+            self.seek_helper(0, key);
+            return;
+        }
+
+        self.seek_helper(idx - 1, key);
+        if IteratorError::check_eof(&self.err) {
+            if idx == self.table.as_ref().offsets_length() {
+                return;
+            }
+            self.seek_helper(idx, key);
+        }
+    }
+
+    // seek_inner will reset iterator and seek to >= key.
+    fn seek_inner(&mut self, key: &Bytes) {
+        self.seek_from(key, SeekPos::Origin);
+    }
+
+    // seek_for_prev will reset iterator and seek to <= key.
+    fn seek_for_prev(&mut self, key: &Bytes) {
+        self.seek_from(key, SeekPos::Origin);
+        if self.key() != key {
+            self.prev_inner();
+        }
+    }
+
+    fn next_inner(&mut self) {
+        self.err = None;
+
+        if self.bpos > self.table.as_ref().offsets_length() {
+            self.err = Some(IteratorError::EOF);
+            return;
+        }
+
+        if BlockIterator::require_load(&self.block_iterator) {
+            match self.table.as_ref().block(self.bpos, self.use_cache()) {
+                Ok(block) => {
+                    let block_iterator = self.get_block_iterator(block);
+                    block_iterator.seek_to_first();
+                    self.err = block_iterator.err.clone();
+                }
+                Err(err) => self.err = Some(err.into()),
+            }
+        } else {
+            let bi = self.block_iterator.as_mut().unwrap();
+            bi.next();
+            if !bi.valid() {
+                self.bpos += 1;
+                bi.data.clear();
+                self.next_inner();
+            }
+        }
+    }
+
+    fn prev_inner(&mut self) {
+        self.err = None;
+
+        if BlockIterator::require_load(&self.block_iterator) {
+            match self.table.as_ref().block(self.bpos, self.use_cache()) {
+                Ok(block) => {
+                    let block_iterator = self.get_block_iterator(block);
+                    block_iterator.seek_to_last();
+                    self.err = block_iterator.err.clone();
+                }
+                Err(err) => self.err = Some(err.into()),
+            }
+        } else {
+            let bi = self.block_iterator.as_mut().unwrap();
+            bi.prev();
+            if !bi.valid() {
+                if self.bpos == 0 {
```

Comment:
You can use `overflowing_sub` to avoid condition check.

Reply:
Now I use `warpping_sub` and check if bpos is `MAX` upon entering `prev_inner`